### PR TITLE
556 Document our approach to uploading data to the API

### PR DESCRIPTION
**Pull request by bell-pa**
## Context

Trello [Document our approach to uploading data to the API](https://trello.com/c/crsvsg0L/556-document-our-approach-to-uploading-data-to-the-api)

## Changes in this PR
* Added a how-to `replace-all-api-data.md`
* Added an ADR `0004-bulk-replace-api-data.md`

This PR does not affect the project's deployable artifacts.

### CODE REVIEW

**Review by Floppy — APPROVED** [review]

File: doc/adr/0004-bulk-replace-api-data.md
Comment (yndajas) [inline doc/adr/0004-bulk-replace-api-data.md:30]: `UI programs` here and below

`HMPPS Auth`

`CSV`

File: doc/adr/0004-bulk-replace-api-data.md
Comment (yndajas) [inline doc/adr/0004-bulk-replace-api-data.md:36]: `IDs` here and elsewhere

File: doc/adr/0004-bulk-replace-api-data.md
Comment (yndajas) [inline doc/adr/0004-bulk-replace-api-data.md:48]: `referrals link to`

File: doc/adr/0004-bulk-replace-api-data.md
Comment (yndajas) [inline doc/adr/0004-bulk-replace-api-data.md:56]: Probably want a line break before this?

`approach requires:`

File: doc/adr/0004-bulk-replace-api-data.md
Comment (yndajas) [inline doc/adr/0004-bulk-replace-api-data.md:53]: `public, and programme`

File: doc/adr/0004-bulk-replace-api-data.md
Comment (yndajas) [inline doc/adr/0004-bulk-replace-api-data.md:59]: `tunneling` and `extracting` (again these points form a sentence based on the structure of the pretext)

`database connection` (no full stop; or you could add a semicolon to the end of the first and retain the full stop, but this would need changing in the previous document/commit too for consistency)

File: doc/adr/0004-bulk-replace-api-data.md
Comment (yndajas) [inline doc/adr/0004-bulk-replace-api-data.md:64]: `bulk upload`

File: doc/adr/0004-bulk-replace-api-data.md
Comment (yndajas) [inline doc/adr/0004-bulk-replace-api-data.md:68]: `Command-line`

File: doc/adr/0004-bulk-replace-api-data.md
Comment (yndajas) [inline doc/adr/0004-bulk-replace-api-data.md:76]: `to support this solution`?

File: doc/adr/0004-bulk-replace-api-data.md
Comment (yndajas) [inline doc/adr/0004-bulk-replace-api-data.md:79]: `solution:` (or make it a `h3`)

**Review by yndajas — COMMENTED** [review]
Apologies for all the comments (proofreading hat on)! This is all very clear, but a few things to tidy up before merging

File: doc/how-to/replace-all-api-data.md
Comment (bell-pa) [reply doc/how-to/replace-all-api-data.md:72]: No. This is how they appeared in the spreadsheet and so this is how they appear in the CSV file.

File: doc/how-to/replace-all-api-data.md
Comment (bell-pa) [reply doc/how-to/replace-all-api-data.md:82]: I don't know if the scripts would work with zsh. I've not tried it. Each script currently has this at the top of the file:
```
#!/usr/bin/env bash
```
I'm not familiar with zsh, but if it is an extension of bash then I suppose it should be OK.
Tech debt?

File: doc/how-to/replace-all-api-data.md
Comment (Gweaton) [inline doc/how-to/replace-all-api-data.md:97]: Typo in the first `cours[e]` here.

**Review by Gweaton — COMMENTED** [review]
This looks good to me, thanks for taking the time to write all of it.

Looks like something strange has happened to the second commit message - possibly a product of squashing, as it's got three lines saying: "Add an Architectural Decision Record which documents the initial approach to uploading course data to the API".

Will approve after tidying that up and the small typo

**Review by Gweaton — APPROVED** [review]

**Pull request merged**
